File: leetcode/cpp/48. 旋转图像.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& m) {
        if (m.empty()) {
            return;
        }

        int n = m.size() - 1;
        for (int k = 0; k <= m.size() / 2; k++) {
            for (int i = 0; i <= n - 2 * k - 1; i++) {
                int tmp = m[k][k + i];
                m[k][k + i] = m[n - k - i][k];
                m[n - k - i][k] = m[n - k][n - k - i]; 
                m[n - k][n - k - i] = m[k + i][n - k];
                m[k + i][n - k] = tmp;
            }
        }

        return;
    }
};

File: leetcode/python/48. 旋转图像.py
#!/bin/bash python
"""
48. 旋转图像

思路:
两层遍历, 进行赋值
"""

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        注意内部不能是 for i in range(left, right)，因此需要的是相对距离
        """
        n = len(matrix)
        top, left, bottom, right = 0, 0, n - 1, n - 1
        while top < bottom and left < right:
            for i in range(right - left):
                tmp = matrix[top][left + i]
                matrix[top][left + i] = matrix[bottom - i][left]
                matrix[bottom - i][left] = matrix[bottom][right - i]
                matrix[bottom][right - i] = matrix[top + i][right]
                matrix[top + i][right] = tmp
            top += 1
            left += 1
            bottom -= 1
            right -= 1
        return